### ISSUE TITLE
Connect external mysql database by egress gateway does not work

### ISSUE THREAD

**Issue by zoupl** [istio/istio#9224]
**I deployed ServiceEntry like this:**

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: cnn
spec:
  hosts:
  - dbext.telcloud.svc.cluster.local
  ports:
  - number: 3306
    name: db
    protocol: TCP
  location: MESH_EXTERNAL
  resolution: STATIC
  endpoints:
  - address: 10.144.166.68
    ports:
      http: 3306

```
**In a pod which have mysql client in it , i tried to connect to mysql and it is ok.**

```
mysql -uroot -p  -h dbext.telcloud.svc.cluster.local -P 3306
Enter password:
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 45
Server version: 5.7.23 MySQL Community Server (GPL)

Copyright (c) 2000, 2018, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql>

```
**Later I deployed egress gateway and virtual service,  then, i tried to connect to mysql,  and it failed.
Here is egress gateway and virtualservice:**

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-egressgateway
spec:
  selector:
    istio: egressgateway
  servers:
  - port:
      number: 3306
      name: db
      protocol: TCP
    hosts:
    - "dbext.telcloud.svc.cluster.local"
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: egressgateway-for-cnn
spec:
  host: istio-egressgateway.istio-system.svc.cluster.local
  subsets:
  - name: cnn


apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: dbext
spec:
  hosts:
  - dbext.telcloud.svc.cluster.local
  gateways:
  - istio-egressgateway
  - mesh
  tcp:
  - match:
    - gateways:
      - mesh
      port: 80
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        subset: cnn
        port:
          number: 80
      weight: 100
  - match:
    - gateways:
      - istio-egressgateway
      port: 80
    route:
    - destination:
        host: dbext.telcloud.svc.cluster.local
        port:
          number: 3306
      weight: 100

```

**Here is my result:**
```
root@mysql-fb75876c6-dcbh8:/# mysql -uroot -p -h dbext.telcloud.svc.cluster.local -P 80
Enter password:
```

**It was stuck, and there was no messages in logs of egress gateway.**

**Expected behavior**
mysql -uroot -p -h dbext.telcloud.svc.cluster.local -P 80, this command should be ok.

**Version**
Istio 1.0.2 release  
kubernetes v1.11.3

**Installation**
Install istio by helm

**Environment**
centos 7.4

**Comment by zoupl** [comment]
I found egress gateway only support 80 and 443, which is http and https. Is TCP not supported by egress gateway?

**Comment by zoupl** [comment]
```
  istio-egressgateway:
    enabled: true
    labels:
      app: istio-egressgateway
      istio: egressgateway
    replicaCount: 1
    autoscaleMin: 1
    autoscaleMax: 5
    cpu:
      targetAverageUtilization: 80
    serviceAnnotations: {}
    type: ClusterIP #change to NodePort or LoadBalancer if need be
    ports:
      - port: 80
        name: http2
      - port: 443
        name: https
    secretVolumes:
      - name: egressgateway-certs
        secretName: istio-egressgateway-certs
        mountPath: /etc/istio/egressgateway-certs
      - name: egressgateway-ca-certs
        secretName: istio-egressgateway-ca-certs
        mountPath: /etc/istio/egressgateway-ca-certs
```

**Comment by supereagle** [comment]
You can refer to [Consuming External TCP Services](https://istio.io/blog/2018/egress-tcp/)

**Comment by crowfrog** [comment]
I meet similar issue with http2 traffic use other port but not 80/443

**Comment by crowfrog** [comment]
hi zoupl,

I think you can change your gateway port from 3306 to 80

apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-egressgateway
spec:
  selector:
    istio: egressgateway
  servers:
  - port:
      number: 80
      name: db
      protocol: TCP
    hosts:
    - "dbext.telcloud.svc.cluster.local"

Then it should work fine.

**Comment by vadimeisenbergibm** [comment]
@zoupl Is your mysql database outside the cluster? I wonder since you use a private IP and a Kubernetes URL.

Regarding your questions - I wrote an example for MongoDB, see this PR https://github.com/istio/istio.io/pull/2347

**Comment by tchanorcl** [comment]
@vadimeisenbergibm @zoupl Hi, I have also encountered the same error - except that it is the Bookinfo->Ratings v2 Node.js throwing an error when it tries to connect to the MySql that is outside the mesh/cluster. MySql is running directly on a host VM. This follows the [Consuming External TCP Services](https://istio.io/blog/2018/egress-tcp/) Except I am trying to redirect all egress traffic to MySQL through the istio-egressgateway with VirtualService, Gateway resources. If I don't try to direct traffic thru istio-egressgateway, and let the Ratings v2 pod use the ServiceEntry directly to the same MySQL db, everything works. 
I also did a tcpdump from the istio-egressgateway pod when it tries to connect to MySQL (proxying the request from Ratings v2 pod), and it shows that the Istio EGateway has somehow got packets out of order and caused MySql to issue error 1156.  So I think this means that Istio Egress Gateway cannot support TCP conns like MySql? 

![image](https://user-images.githubusercontent.com/45220231/49532130-d4d90300-f889-11e8-83d1-5eb4ed659ecc.png)

**Comment by zoupl** [comment]
Istio egress gateway still not work. So I changed to use haproxy instead.

**Comment by vadimeisenbergibm** [comment]
@tchanorcl Could you please see https://preliminary.istio.io/blog/2018/egress-mongo/? It describes multiple options for directing TCP/TLS egress traffic thru the egress gateway.

**Comment by tchanorcl** [comment]
@vadimeisenbergibm - Hi, thank you for responding. I tried the https://preliminary.istio.io/blog/2018/egress-mongo/ instructions (Direct TCP Egress traffic through an egress gateway) and it _**WORKS!**_. The mods I made were for MySQL port/host. However, the ServiceEntry apply failed initially because of "Error from server: error when creating "vad_mysql_egress_service_entry.yaml": admission webhook "pilot.validation.istio.io" denied the request: configuration is invalid: **CIDR addresses are allowed only for NONE/STATIC resolution types**" - As per instructions on the prelim blog, the "resolution" property is set to "DNS" and included the "addresses" property, but Istio (for me, Version 1.0.4) didn't like specifying the "addresses" property if set to DNS, So I commented out the "addresses" and redid the apply. 

Log from Ratings v2 Pod: Shows it is going to the egressgateway (Great!)
[2018-12-06T15:44:27.555Z] - 107 180 9 "10.201.2.219:7777" outbound|7777|mysql|istio-egressgateway.istio-system.svc.cluster.local 10.201.0.137:60772 10.0.20.6:3306 10.201.0.137:40474

Log from the Egress Gateway Pod: Shows it is proxying the connection to MySql (Working as expected)
[2018-12-06T15:44:26.861Z] - 107 180 25 "10.0.20.6:3306" outbound|3306||chantclientphx101.publicetcdad1.k8soci.oraclevcn.com 10.201.2.219:43490 10.201.2.219:7777 10.201.0.137:60766

Again, thanks for the feedback and instructions.
